### ISSUE TITLE
Consider a "file per test case" testing structure

### ISSUE THREAD

**Issue by patrickpilch** [arturbosch/detekt#1828]
In the course of developing a rule augmentation, I struggled with writing tests to my intention. I wanted to detail some of the difficulties I've faced along with my opinions, and propose a new test case structure. I'm certain there are good reasons for the existing structure that I do not have insight into due to my lack of the project's historical context, but I'm hoping that my suggestions elicit improvements that address most of them.

The current structure of maintaining one file for Positive & Negative cases respectively makes it more challenging (IMO) to write a test case that is maintainable over the longer course of an actively developed rule. My expectation of a rule test case is that it ensures a _particular message_ is reported for a _particular unit of code._ Information such as class, line, containing method, etc. are cross-cutting concerns and are out of the scope of a rule test (and instead would live within the issue reporting functionality test cases.)

So far, I see a few approaches that I consider not ideal in that regard:
### Multiple test cases in one `Case`
Some test cases run `lint` over the entirety of the Positive file, and assert the expected quantity of issues reported. This has the disadvantages of test failure traceability. Consider the scenario where I've modified a rule, and the 4th out of the 10 total positive cases is now incorrectly not reported. The test output indicates an expectation of "10" but is now "9", and spews out a list of 9 `Issue` string representations containing long file paths, line numbers, etc. that you now have to sort out. 

Placing all the cases into one file also increases the difficulty of asserting error messaging. Consider the example above, but now with also having assertions specifying that issue 4/10 should have a distinct message from the others. The reporting will now be off by one position. 
This issue will also affect you when test cases are _added_ in between existing ones (for organization's sake). Using line numbers in the assertions for a specific rule test faces all of the same issues, where merely adding a line to a large Case will result in tons of assertion changes (feels like a violation of the open/closed principle to me.)

### The `Case` enum
This enum provides a mapping of test cases to file paths along with some utility functions. But because it's an enum, it dictates a strictly one-level hierarchy in test case organization. This means that if I wanted to maintain multiple test case files, they would all have to be in the same directory and prefixed so they are grouped with related cases. This can result in some lengthy names like `SomeRuleSomeQualifierSomeSpecificInstancePositive`. 

## Proposal
I propose to use the Java packaging system as the organizational system for the `Rule`s and their test cases. If each rule has its own package, there will be a corresponding package inside the test `resources` folder that can contain multiple files and a directory structure. I think this gives the test designer more flexibility in how to group, organize and write the tests.

One can add another extension function to `RuleExtensions` like
```kotlin
inline fun <reified T : Rule> T.lintCase(caseName: String, clazz: Class<T> = this.javaClass): List<Finding> {
    val content = clazz.getResource(caseName).readText()
    return this.compileAndLint(content)
}
```
The unit tests can then read more like so:
```kotlin
describe("Something rule") {
    context("some exemption flag set to true") {
        it("reports an incorrect usage of X") {
            val issues = subject.lintCase("case_x.kt")
            assertThat(issues).hasSize(1)
            assertThat(issues[0].message).equals("Can not use X when flag is set to true")
        }
    }
    context("some exeption flag set to false") {
        it("does not report usage of X") {
            // ...
        }
    }
}
```
In this example, re-use of the same Case code is possible for testing whether it yields an issue or not based on arguments set on the rule. And testing for messages is less brittle, where adding more functionality & cases will not force you to touch the existing cases or assertions. There is also the added flexibility of creating sub-packages to separate out the cases for enabled/disabled, etc.

There's even the possibility of taking it a step further and creating a custom AssertJ Matcher to write friendlier looking assertions (perhaps removing the boilerplate around asserting size of 1, grabbing the index 0, doing equality checks on strings, etc.) but perhaps that's a stretch :)

**Comment by 3flex** [comment]
Thanks for the detailed proposal! You may not have seen, but not all test cases use the problematic style - some embed the code to be tested within the test case, which is great since it keeps the entire test case self contained and there are no references to external files. They typically only expect a single error so failures are easier to check too.

Personally I'd like to see things move in that direction since it's easier to review and has less boilerplate.

Your suggestion would work well for test cases that require multiple files. I don't know off hand if we have any today, but sample code for that case would be good to see in your proposal.

**Comment by schalkms** [comment]
Hi Patrick! Thanks for your detailed proposal, which brings new ideas to detekt's testing approach. We have actually thought about this in issue #1089. So I'll comment your thoughts and we will continue discussing in the other issue, if that's fine for you. 🙂

> The current structure of maintaining one file for Positive & Negative cases respectively makes it more challenging (IMO) to write a test case that is maintainable over the longer course of an actively developed rule.

Yes, you are right. We are in the transformation phase from the positive/negative file approach to having tests that exactly assert 1 violation in 1 test case. This requires us to write test snippets embedded in Kotlin strings. However, the problem here is that you could easily write Kotlin code, which is invalid and doesn't compile at all. In order to prevent this from happening, detekt compiles and thus verifies test snippets on the fly when executing the tests.

> My expectation of a rule test case is that it ensures a particular message is reported for a particular unit of code. Information such as class, line, containing method, etc. are cross-cutting concerns and are out of the scope of a rule test (and instead would live within the issue reporting functionality test cases.)

I agree with the first sentence. We only check for the right message to be reported, when there's more than one report() statement in the respective rule. Checking for the mentioned information is therefore not always a cross-cutting concern, but of course we also have tests for the *generic* reporting logic.

> Multiple test cases in one Case

I agree with all the arguments in this paragraph. This has been discussed intensively in #1089. That's the reason for migrating away from this approach. We could certainly use help here.
I did some work on the migration, but I couldn't complete it due to time constraints. Furthermore, there were also more important tasks to do on the way to version 1.0.0.
Working on the migration is unfortunately a bit of a tedious task.

> This enum provides a mapping of test cases to file paths along with some utility functions. But because it's an enum, it dictates a strictly one-level hierarchy in test case organization. 

With the new approach we cover hierarchies in the test case organization using the Spek testing framework.

> Proposal

I'll link a PR in a moment that shows how we write tests cases for our rule.
@patrickpilch maybe you have time to look at this PR and comment your thoughts. I would be very interested in discussing this with you.

------------

It's nice to hear that you write custom rules. Please feel free to submit a PR, if you have rules that could be beneficial for other detekt users. 🙂

I also want to note that detekt doesn't force you how to write tests for your domain-specific rules. Detekt just provides an infrastructure, which helps you to complete your task.

**Comment by schalkms** [comment]
Discussion moved to #1089
